optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/Duration.java

```diff
@@ -295,67 +295,56 @@ public static Duration between(DateTimeAccessor startInclusive, DateTimeAccessor
    * @throws DateTimeParseException if the text cannot be parsed to a {@code Duration}
    */
   public static Duration parse(final CharSequence text) {
-
     Jdk7Methods.Objects_requireNonNull(text, "text");
-    int len = text.length();
-    if (len < 4 || (text.charAt(0) != 'P' && text.charAt(0) != 'p') || (text.charAt(1) != 'T' && text.charAt(1) != 't')
-        || (text.charAt(len - 1) != 'S' && text.charAt(len - 1) != 's')
-        || (len == 5 && text.charAt(2) == '-' && text.charAt(3) == '0')) {
-      throw new DateTimeParseException("Duration could not be parsed: " + text, text, 0);
-    }
-    String numberText = text.subSequence(2, len - 1).toString().replace(',', '.');
-    if (numberText.charAt(0) == '+') {
-      throw new DateTimeParseException("Duration could not be parsed: " + text, text, 2);
-    }
-    int dot = numberText.indexOf('.');
+    //PTnHnMnS,
     try {
-      if (dot == -1) {
-        // no decimal places
-        if (numberText.startsWith("-0")) {
-          throw new DateTimeParseException("Duration could not be parsed: " + text, text, 2);
+        if (Character.toUpperCase(text.charAt(0)) != 'P' && Character.toUpperCase(text.charAt(1)) != 'T') {
+            throw new DateTimeParseException("Duration could not be parsed: " + text, text, 0); 
         }
-        return create(Long.parseLong(numberText), 0);
-      }
-      // decimal places
-      boolean negative = false;
-      if (numberText.charAt(0) == '-') {
-        negative = true;
-      }
-      long secs = Long.parseLong(numberText.substring(0, dot));
-      numberText = numberText.substring(dot + 1);
-      len = numberText.length();
-      if (len == 0 || len > 9 || numberText.charAt(0) == '-' || numberText.charAt(0) == '+') {
-        throw new DateTimeParseException("Duration could not be parsed: " + text, text, 2);
-      }
-      int nanos = Integer.parseInt(numberText);
-      switch (len) {
-        case 1:
-          nanos *= 100000000;
-          break;
-        case 2:
-          nanos *= 10000000;
-          break;
-        case 3:
-          nanos *= 1000000;
-          break;
-        case 4:
-          nanos *= 100000;
-          break;
-        case 5:
-          nanos *= 10000;
-          break;
-        case 6:
-          nanos *= 1000;
-          break;
-        case 7:
-          nanos *= 100;
-          break;
-        case 8:
-          nanos *= 10;
-          break;
-      }
-      return negative ? ofSeconds(secs, -nanos) : create(secs, nanos);
-
+        Duration out = Duration.ZERO;
+        StringBuilder number = new StringBuilder();
+        int state = 0;
+        for (int i = 2; i < text.length(); i++) {
```

Comment:
Where do these changes come from?

Reply:
`java.time.Duration` toString() (sent by server) may have hours and minutes, the old implementation only accepted seconds